File: 5_abstractDataStructures/4_binary_trees_n_binary_search_tree/2_tree_breadth_first_search/3_zigzag_tree_traversal.js
/*

Problem Statement #

Given a binary tree, populate an array to represent its zigzag level order traversal. 
You should populate the values of all nodes of the first level from left to right, 
then right to left for the next level and keep alternating in the same manner for the following levels.


Solution #

This problem follows the Binary Tree Level Order Traversal pattern. We can follow the same BFS approach. 
The only additional step we have to keep in mind is to alternate the level order traversal, 
which means that for every other level, we will traverse similar to Reverse Level Order Traversal question.

*/

class Node {
    constructor(value) {
        this.value = value;
        this.left = null;
        this.right = null;
    }
}

class BinaryTree {
    constructor(root = null) {
        this.root = root;
    }

    zigzap_tree_traversal(){
        let result = [];

        let currentNode = this.root;

        if(currentNode == null){
            return result;
        }

        let depthCounter = 0;  // tracks what order it should take...left to right(for odd levels) or right to left(for even levels)

        let queue = [];
        queue.push(currentNode);

        while(queue.length >= 1){
            let levelNodesCollector = [];
            let queueLenghtSnapshot = queue.length;

            for(let i = 0; i < queueLenghtSnapshot; i++) {
                let node = queue.shift();
                
                // if even, then normal order
                if(depthCounter % 2 == 0){
                    levelNodesCollector.push(node);
                }
                // if odd, then reverse order
                else{
                    levelNodesCollector.unshift(node)
                }

                if(node.left != null){
                    queue.push(node.left);
                }
                if(node.right != null){
                    queue.push(node.right);
                }
            }

            result.push(levelNodesCollector);
            depthCounter++;
        }

        return result;
    }
}


//         10
//     5       15
//  2    8  12   18



let root = new Node(10);
root.left = new Node(5);
root.right = new Node(15);
root.left.left = new Node(2);
root.left.right = new Node(8);
root.right.left = new Node(12);
root.right.right = new Node(18);

const bt = new BinaryTree(root);
const res = bt.zigzap_tree_traversal();
console.log("Result2: ", res);

File: 5_abstractDataStructures/4_binary_trees_n_binary_search_tree/2_tree_breadth_first_search/3_zigzag_tree_traversal.java
// NB: To run, class name must match file name

import java.util.*;

class Node {
    int value;
    Node left, right;

    // Constructor to initialize the node
    public Node(int value) {
        this.value = value;
        this.left = null;
        this.right = null;
    }
}

class BinaryTree {
    Node root;

    // Constructor to initialize the binary tree
    public BinaryTree(Node root) {
        this.root = root;
    }

    public List<List<Integer>> zigzagTreeTraversal() {
        List<List<Integer>> result = new ArrayList<>();

        if (root == null) {
            return result; // If the tree is empty, return empty result
        }

        Queue<Node> queue = new LinkedList<>();
        queue.offer(root);

        boolean leftToRight = true; // Tracks the direction of traversal
        
        while (!queue.isEmpty()) {

            int levelSize = queue.size();
            List<Integer> levelNodes = new ArrayList<>();

            for (int i = 0; i < levelSize; i++) {
                Node node = queue.poll();

                // If we are traversing left to right, add the node normally at END
                if (leftToRight) {
                    levelNodes.add(node.value);
                } else {
                    // If we are traversing right to left, add the node at the FRONT
                    levelNodes.add(0, node.value);
                }

                // Add child nodes to the queue
                if (node.left != null) {
                    queue.offer(node.left);
                }
                if (node.right != null) {
                    queue.offer(node.right);
                }
            }

            // Add the current level's result to the final list
            result.add(levelNodes);

            // Toggle the direction for the next level
            leftToRight = !leftToRight;
        }

        return result;
    }

    public static void main(String[] args) {
        // Create the binary tree
        Node root = new Node(10);
        root.left = new Node(5);
        root.right = new Node(15);
        root.left.left = new Node(2);
        root.left.right = new Node(8);
        root.right.left = new Node(12);
        root.right.right = new Node(18);

        // Initialize the BinaryTree object
        BinaryTree bt = new BinaryTree(root);

        // Get the zigzag traversal result
        List<List<Integer>> result = bt.zigzagTreeTraversal();

        // Print the result
        System.out.println("Zigzag Level Order Traversal:");
        for (List<Integer> level : result) {
            System.out.println(level);
        }
    }
}